hash_table.h

```diff
@@ -1,57 +1,95 @@
-// pass one thing on insert()?
+    HashTable() = delete;
 
-// Usage notes:
-// 1) A pointer should always be passed as T.
-// 2) It's the user's responsibility to not insert duplicate `id`.
+    explicit HashTable(size_t buckets)
+        : size{buckets}
+        , ids{static_cast<const char **>(allocate_zero(buckets, mem))}
+        , data{static_cast<T *>(allocate_zero(buckets, mem))}
+    {
+        for (size_t i = 0U; i != size; ++i) {
+            ids[i] = nullptr;
+            data[i] = nullptr;
+        }
+    }
 
-template<typename T>
-struct HashTable {
-    static_assert(std::is_pointer<T>::value, "T must be pointer");
+    HashTable(const HashTable<T, mem, H> &rhs) = delete;
+    HashTable &operator=(const HashTable<T, mem, H> &rhs) = delete;
+    HashTable(HashTable<T, mem, H> &&rhs) = delete;
+    HashTable &operator=(HashTable<T, mem, H> &&rhs) = delete;
 
-    Buf<const char *> ids;
-    Buf<T> data;
+public:
 
-    HashTable() { }
+    inline void insert(const char *key, T value) {
+        size_t bucket = hasher(const_cast<char *>(key)) % size;
+        size_t i = bucket;
+        bool wrapped = false;
+        while (!wrapped) {
+            if (ids[i] == nullptr) {
+                ids[i] = key;
+                data[i] = value;
+                return;
+            } else if (ids[i] == key) {
+                assert(false);
+            }
+            i = (i + 1U) % size;
+            wrapped = (i == bucket); 
+        }
 
-    HashTable(size_t n) {
-        this->reserve(n);
+        assert(false);
     }
 
-    void reserve(size_t n) {
-        ids.reserve(n);
-        data.reserve(n);
-    }
+    inline T find(const char *key) {
+        size_t bucket = hasher(key) % size;
+        size_t i = bucket;
+        bool wrapped = false;
+        while (!wrapped) {
+            if (ids[i] == key) { 
+                return data[i];
+            } else if (ids[i] == nullptr) {
+                return nullptr;
+            }
 
-    void insert(const char *id, T v) {
-        assert(this->find(id) == NULL);
-        ids.push(id);
-        data.push(v);
+            i = (i + 1U) % size;
+            wrapped = (i == bucket);
+        }
+        return nullptr;
     }
 
-    T find(const char *id) {
-        size_t i = 0;
-        for (const char *ident : ids) {
-            if (ident == id) {
-                return data[i];
+    inline bool contains_key(const char *key) {
+        size_t bucket = hasher(key) % size;
+        size_t i = bucket;
+        bool wrapped = false;
+        while (!wrapped) {
+            if (ids[i] == key) {
+                return true;
+            } else if (ids[i] == nullptr) {
+                return false;
             }
-            ++i;
+            i = (i + 1U) % size;
+            wrapped = (i == bucket);
         }
-        // Can do that because we have verified that T is a pointer.
-        return nullptr;
+        return false;
     }
+
+private:
+    const char **ids;
+    size_t size;
```

Comment:
This should be `cap` or something (denoting capacity and not the length / size, i.e. num of items). Then, maybe we need another field that denotes that.